services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/Hakemusalue.kt

```diff
@@ -60,17 +65,99 @@ data class KaivuilmoitusAlue(
+        checkChange(index, KaivuilmoitusAlue::tyonTarkoitukset, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::meluhaitta, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::polyhaitta, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::tarinahaitta, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::kaistahaitta, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::kaistahaittojenPituus, other)?.let { changes.add(it) }
+        checkChange(index, KaivuilmoitusAlue::lisatiedot, other)?.let { changes.add(it) }
+        // haittojenhallintasuunnitelma changes do not affect the area itself, therefore we add the
+        // root area here if there are changes in other fields.
+        if (changes.isNotEmpty()) {
+            changes.add(0, "areas[$index]")
+        }
+        changes.addAll(
+            checkChangesInHaittojenhallintasuunnitelma(
+                index,
+                haittojenhallintasuunnitelma,
+                other.haittojenhallintasuunnitelma,
+            )
+        )
+        return if (changes.isNotEmpty()) changes else null
+    }
 }
 
 data class Tyoalue(
     val geometry: Polygon,
     val area: Double,
     val tormaystarkasteluTulos: TormaystarkasteluTulos?,
-)
+) {
+    fun listChanges(alueIndex: Int, tyoalueIndex: Int, other: Tyoalue?): List<String>? {
+        if (other == null) return listOf("areas[$alueIndex].tyoalueet[$tyoalueIndex]")
+        val changes = mutableListOf<String>()
+        checkChange(alueIndex, tyoalueIndex, Tyoalue::geometry, other)?.let { changes.add(it) }
+        return if (changes.isNotEmpty())
+            listOf("areas[$alueIndex].tyoalueet[$tyoalueIndex]") + changes
+        else null
+    }
+}
 
 fun List<KaivuilmoitusAlue>?.combinedAddress(): PostalAddress? =
     this?.map { it.katuosoite }
         ?.toSet()
         ?.joinToString(", ")
         ?.let { StreetAddress(it) }
         ?.let { PostalAddress(it, "", "") }
+
+private fun KaivuilmoitusAlue.checkChange(
+    index: Int,
+    property: KProperty1<KaivuilmoitusAlue, Any?>,
+    second: KaivuilmoitusAlue,
+): String? =
+    if (property.get(this) != property.get(second)) {
+        "areas[$index].${property.name}"
+    } else null
+
+private fun Tyoalue.checkChange(
+    alueIndex: Int,
+    tyoalueIndex: Int,
+    property: KProperty1<Tyoalue, Any?>,
+    second: Tyoalue,
+): String? =
+    if (property.get(this) != property.get(second)) {
+        "areas[$alueIndex].tyoalueet[$tyoalueIndex].${property.name}"
+    } else null
+
+private fun checkChangesInTyoalueet(
+    index: Int,
+    firstAreas: List<Tyoalue>,
+    secondAreas: List<Tyoalue>,
+): List<String> {
+    val changedElementsInFirst =
+        firstAreas.withIndex().mapNotNull { (i, tyoalue) ->
+            tyoalue.listChanges(index, i, secondAreas.getOrNull(i))
```

Comment:
Since there's only one field that can be different in a tyoalue, it could be checked right here.

Reply:
I moved the check for `Tyoalue.geometry` here.